### ISSUE TITLE
progress bar indicates how many iterations are complete, or which iteration is being worked on?

### ISSUE THREAD

**Issue by abw333** [tqdm/tqdm#306]
When I run the following code on Python 2.7:

```
import time
import tqdm

def test():
    for x in range(100): yield x
    time.sleep(1)
    for x in range(100): yield x
    time.sleep(1)

for x in tqdm.tqdm(test(), total=200):
    pass
```

the progress bar stops at 101 for a second, and then at 200.

When it stops at 101, it would seem to be indicating that 100 iterations are complete, and that the 101st is being worked on. When it stops at 200, it's indicating that 200 iterations are complete.

Is this behavior expected? If so, it seems inconsistent. What's the rationale behind it?

(This is a minimal reproduction of the issue. The actual use case consists of splitting a large payload into various batched requests for a web API, and yielding the results after each API call.)

**Comment by lrq3000** [comment]
Yes this is the expected behavior. There's not really a rationale but we want to show what iteration we are working on, so when the bar is initialized it shows the first iteration (instead of 0), and when it's done the bar closes with the last state, which in your case is 200. So at the last iteration, the number of iterations does not change, the bar simply closes.

I can't see a way around this except if we start at 0, and show only the number of iterations done, not the current iteration.

But technically, this would be a big change in the internals of tqdm, because we print the progress bar always after the item gets yield. If we start at 0, we would first print tqdm and then yield, and then close the bar if necessary after the last item is yield. It's possible, but it's really another approach.

**Comment by abw333** [comment]
I see, thanks for the quick response!

One follow up question: You mentioned that the progress bar starts on the first iteration instead of 0. When I run the code above, I actually see 0, then 101. Interestingly, if I had another `time.sleep(1)` before the first `for x in range(100): yield x`, I see 0, then 1, then 101. Is that expected behavior?

**Comment by lrq3000** [comment]
Yeah i was simplifying sorry, in the past it used to work exactly as i
described becaus tqdm was only a function, but now it's a class so it
prints the 0 state at initialization. But then the bar gets only updated
after x is yielded.

**Comment by abw333** [comment]
OK great, thanks!

**Comment by casperdcl** [comment]
> I can't see a way around this except if we start at 0, and show only the number of iterations done, not the current iteration.

I'd prefer that to the current implementation as it's more standard and intuitive. progress = tasks finished, not tasks running+finished.
